### PR TITLE
Use separate convolutions in ResizeCPTransform

### LINKED ISSUE #722: Use separable kernels for ResizeCPTransform
## Subject of the feature

Currently, in `ResizeCPTransform` we are using 3D Gaussain kernel https://github.com/DeepRegNet/DeepReg/blob/main/deepreg/model/layer.py#L435

I think the 1D separable kernel is equivalent to this and it's more efficient? As in LNCC https://github.com/DeepRegNet/DeepReg/blob/main/deepreg/loss/image.py#L182

So I suggest to uniform this bit, what do you think @acasamitjana

### PR DESCRIPTION
**Pull request by acasamitjana**
# Description

Use separable convolutions to speed up the low pass filtering in ResizeCPTransform

Fixes #722 

## Type of change

What types of changes does your code introduce to DeepReg?

_Please check the boxes that apply after submitting the pull request._

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, renaming)
- [X] Refactoring (no functional changes, no api changes)
- [ ] Documentation Update (fix or improvement on the documentation)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (if none of the other choices apply)

## Checklist

_Please check the boxes that apply after submitting the pull request._

_If you're unsure about any of them, don't hesitate to ask. We're here to help! This is
simply a reminder of what we are going to look for before merging your code._

- [X] I have
      [installed pre-commit](https://deepreg.readthedocs.io/en/latest/contributing/setup.html)
      using `pre-commit install` and formatted all changed files. If you are not
      certain, run `pre-commit run --all-files`.
- [X] My commits' message styles matches
      [our requested structure](https://deepreg.readthedocs.io/en/latest/contributing/commit.html),
      e.g. `Issue #<issue number>: detailed message`.
- [X] I have updated the
      [change log file](https://github.com/DeepRegNet/DeepReg/blob/main/CHANGELOG.md)
      regarding my changes.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by acasamitjana** [comment]
Hi @mathpluscode @YipengHu can you check that?

File: deepreg/model/layer.py [hunk 1]
```diff
@@ -425,25 +427,31 @@ def __init__(
...
         self._resize = None
 
     def build(self, input_shape):
         super().build(input_shape=input_shape)
 
-        self.kernel = layer_util.gaussian_filter_3d(self.kernel_sigma)
         output_shape = tuple(
-            tf.cast(tf.math.ceil(v / c) + 3, tf.int32)
-            for v, c in zip(input_shape[1:-1], self.cp_spacing)
+            [
```
Comment (mathpluscode) [inline deepreg/model/layer.py:438]: no need of `[ ]` here, `tuple(x for x in xx)` is fine, so this line does not need to be modified

File: deepreg/model/layer.py [hunk 2]
```diff
@@ -425,25 +427,31 @@ def __init__(
...
+                tf.cast(tf.math.ceil(v / c) + 3, tf.int32)
+                for v, c in zip(input_shape[1:-1], self.cp_spacing)
+            ]
         )
         self._output_shape = output_shape
         self._resize = Resize3d(output_shape)
 
     def call(self, inputs, **kwargs) -> tf.Tensor:
-        output = tf.nn.conv3d(
```
Comment (mathpluscode) [inline deepreg/model/layer.py:444]: I thought we just need to replace `conv3d` by `separable_filter` ? why you got an extra for loop?

File: deepreg/model/layer_util.py [hunk 3]
```diff
@@ -369,63 +368,6 @@ def warp_grid(grid: tf.Tensor, theta: tf.Tensor) -> tf.Tensor:
     return grid_warped
 
 
-def gaussian_filter_3d(kernel_sigma: Union[Tuple, List]) -> tf.Tensor:
```
Comment (mathpluscode) [inline deepreg/model/layer_util.py:372]: I would keep this func and use it for unit-testing, as the separable conv should be identical to the conv3d option? can we test like this?

File: deepreg/model/layer.py (hunk 2 above)
Comment (acasamitjana) [reply deepreg/model/layer.py:444]: Our current `separable_filter` implementation assumes that channel dimension = 1. Since the filter has to be applied independently to each channel, I use the for loop to apply the filter to each channel.

Comment (mathpluscode) [reply deepreg/model/layer.py:444]: I see. Let me have another look later :)
